### DISCUSSION TITLE
libicucore.dylib location on BigSur

### DISCUSSION THREAD

**Discussion by LecrisUT** [Homebrew/discussions#1121] [Everyday usage] [answered]
According to `icu4c`, `macOS provides libicucore.dylib (but nothing else)`, and if I do a search for it I find it in XCode.app SDKs, but not in /usr/lib or similar spots. I'm trying to create a formula for `webkitgtk`, but I'm running in a snag with cmake `No rule to make target `/usr/lib/libicucore.dylib'`. Does anyone have experience with this recently?

EDIT: More details on this issue: I see that all of the `libicucore.dylib` files provided there are for non MacOS. In the MacOS's runtime folder it is `libicucore.tbd`

**Comment by Bo98** [comment, marked as answer]
Big Sur doesn't provide on-disk dylibs - system dylibs are now stored in the dyld cache.

You can check it via `dlopen` call, such as: `ruby -rfiddle -e "Fiddle.dlopen('/usr/lib/libicucore.dylib')"`

There is probably a problem with the CMake script. The Big Sur changes did break a few build scripts that did some unusual things.

**Reply by LecrisUT** [reply]
Thanks for the tips. I tried the ruby script but it didn't print anything. How should the CMake find the dyld cache in these situations?

**Reply by Bo98** [reply]
Ah yes, I should clarify that no output = successfully found. The Ruby line will print an error if it can't find the library.

> How should the CMake find the dyld cache in these situations?

Depends a bit on what it's doing in the first case. In general, if you link against `icucore` then you don't need to do anything else - I suspect your existing build script is not doing something as simple as that though.

**Reply by LecrisUT** [reply]
Yes the source is from webkit and I'm setting up the IDE to investigate deeper. From the debug log, I can see that the ICU dependence comes from `harfbuzz`
```
-- Found the following HarfBuzz libraries:
--  HarfBuzz (required): /usr/local/lib/libharfbuzz.dylib
--  ICU (required): /usr/local/lib/libharfbuzz-icu.dylib
-- Found HarfBuzz: /usr/local/include/harfbuzz (found suitable version "2.8.0", minimum required is "0.9.18") 
```
I'm not seeing other direct `FindPackage(icu)` in the log, but I'll check the code as well.

PS: unrelated, but are there any IDE setups for homebrew development? I would like to setup a CLion dev environment to more easily debug these stuff.

**Reply by SMillerDev** [reply]
> PS: unrelated, but are there any IDE setups for homebrew development? I would like to setup a CLion dev environment to more easily debug these stuff.

Homebrew is written in ruby so you're better off with RubyMine. If you want to debug all software in homebrew you'll need just about every IDE out there.

**Reply by LecrisUT** [reply]
@SMillerDev yes, for the formula itself and other core Homebrew dev that would help. I think the more relevant part is: How to clear the system's environment and use brew's shell instead. One option I can think of is to open the IDE within a homebrew shell, but I have no luck doing so yet.

**Reply by SMillerDev** [reply]
I'm not sure why you need to clear anything on the system or use the homebrew shell. The only way to test a formula you write is trying to install it. So for me that means:
1. `brew create` to create a formula
1. Fill in the steps from the upstream install guide into the `install` block.
1. Run `brew install --build-from-source --debug formula_name` to see if it works.
1. When it works, add test block and run `brew test formula_name` until it works.

**Reply by LecrisUT** [reply]
Yes, those work for building the formula from cli, but in order to find a problematic section of the code to patch, e.g. this situation, it would be more convenient to use an IDE and navigate the code there. The problem is the IDE would usually pick up non-homebrew installation as well, e.g. an intel compiler, and is not as bare-bone as the brew shell which is more adequate for formula creation and debugging. Ideally if there is a command to reset the current sell env and make  one similar to the ones in the `brew install --build-from-source` stage, that can be integrated in the IDE.

**Reply by SMillerDev** [reply]
The brew shell environment in an install is dependent on what you depend on in the formula. You can try your luck with `brew sh` but in the end the only way to test an install is to try installing something.

As for patching: https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff might be worth reading as well.

**Reply by LecrisUT** [reply]
So back to the topic, I have the IDE up and running to debug (not ideal configuration, just hacked up for now). The specific error is:
```
make[2]: *** No rule to make target `/usr/lib/libicucore.dylib', needed by `bin/LLIntSettingsExtractor'.  Stop.
make[1]: *** [Source/JavaScriptCore/CMakeFiles/LLIntSettingsExtractor.dir/all] Error 2
```
And it seems the target is automatically generated into `make all`, but I cannot find the CMake section that generates it. In the source code all I can find related to it is:
```
add_executable(LLIntSettingsExtractor
    ${JAVASCRIPTCORE_DIR}/llint/LLIntSettingsExtractor.cpp
    ${JavaScriptCore_DERIVED_SOURCES_DIR}/LLIntDesiredSettings.h ${JavaScriptCore_DERIVED_SOURCES_DIR}/Bytecodes.h ${JavaScriptCore_DERIVED_SOURCES_DIR}/BytecodeStructs.h
)
target_link_libraries(LLIntSettingsExtractor WebKit::WTF)
add_dependencies(LLIntSettingsExtractor JavaScriptCore_CopyHeaders)
```
Everything seems to build correctly, just at the linking stage it fails. I think there it comes from one of the inner components, e.g. in the `WTF` there is:
```
list(APPEND WTF_LIBRARIES
    ICU::data
    ICU::i18n
    ICU::uc
)
```
But here I'm drawing blank I can't find the relevant package for that to see how `/usr/lib/libicucore.dylib` target is created.

**Reply by SMillerDev** [reply]
This is all very specific to the software you're trying to build though. Even though we package a lot of software and might be able to help you debug this, it's much easier to contact the people that wrote the software and have them explain what created this target.

I'm also starting to wonder, does this software work on macOS outside of brew? Because I don't believe we do anything special for system libraries.

**Reply by LecrisUT** [reply]
Yes it is specific to there and I've filed a bug report for that to see if they can point me to some resolution.

> I'm also starting to wonder, does this software work on macOS outside of brew?

Yes, there is a GTK version for MacPorts and non-GTK version native for MacOS. Also there are past formulae in other taps which worked for older MacOS and older `webkit`. However, the macports version is not actively maintained and also is not built on BigSur. For the non-GTK version, should I try make a cask and see if the `pdfpc` formula (for which I'm building this formula for) will be able to be picked-up?

But if anyone had experience with how to link those libraries in cmake, I should be able to patch it and check if the formula works otherwise.

**Reply by carlocab** [reply]
Have you tried using the `icu4c` formula? That actually has `data`, `i18n`, and `uc` libraries that `WTF` seems to be looking for.

**Reply by LecrisUT** [reply]
> Have you tried using the `icu4c` formula? That actually has `data`, `i18n`, and `uc` libraries that `WTF` seems to be looking for.

Yes, partly. I have it in the `depends_on`, but didn't change any env variables to make it specifically compatible. I will double check with how to specifically link to it and see if it works any better.

**Reply by LecrisUT** [reply]
@carlocab Progress! I've setup `ICU_ROOT=/usr/local/opt/icu4c` manually and it seems to have overcome that issue. So question is why homebrew does not pick it up on `depends_on`, and if I need to set it up manually, what's the proper way to do so in the formula?

**Reply by SMillerDev** [reply]
> So question is why homebrew does not pick it up on `depends_on`, and if I need to set it up manually, what's the proper way to do so in the formula?

Because homebrew doesn't really know what variables to use. But if there is an `ICU_ROOT` you can use the way the QT formula does it: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/qt.rb#L89

**Reply by LecrisUT** [reply]
Gotcha, thanks for the help as always @SMillerDev .
